crates/ra_ide/src/inlay_hints.rs

```diff
@@ -95,6 +94,15 @@ fn get_chaining_hints(
     let next = tokens.next()?.kind();
     let next_next = tokens.next()?.kind();
     if next == SyntaxKind::WHITESPACE && next_next == SyntaxKind::DOT {
+        let ty = sema.type_of_expr(&expr)?;
+        if ty.is_unknown() {
+            return None;
+        }
+        if let Some(Adt::Struct(st)) = ty.as_adt() {
+            if st.fields(sema.db).is_empty() {
```

Comment:
I'm still not a fan of the extra allocations.

Reply:
You cannot avoid them completely with the current approach to the API, where we avoid excessive reference and lifetimes usage, so I think it's a compromise we have to live with for now.

Reply:
Why not making this a getter like `kind -> Union | Record | Tuple`. But, here is what's thought. This will be a false-positive when the expression is not the struct literal, but e.g. some function that returns a unit or empty-tuple struct. Maybe we lack `ra_hir::Expr` representation for such cases where we want to determine the kind of expression?

Reply:
Good point, added a check for `PathExpr`.